src/service-worker.ts

```diff
@@ -0,0 +1,122 @@
+// for the list of available Workbox modules, or add any other
+// code you'd like.
+// You can also remove this file if you'd prefer not to use a
+// service worker, and the Workbox build step will be skipped.
+
+import { clientsClaim } from 'workbox-core';
+import { ExpirationPlugin } from 'workbox-expiration';
+import { createHandlerBoundToURL, precacheAndRoute } from 'workbox-precaching';
+import { registerRoute } from 'workbox-routing';
+import { CacheFirst, StaleWhileRevalidate } from 'workbox-strategies';
+
+declare const self: ServiceWorkerGlobalScope;
+
+clientsClaim();
+
+// Precache all of the assets generated by your build process.
+// Their URLs are injected into the manifest variable below.
+// This variable must be present somewhere in your service worker file,
+// even if you decide not to use precaching. See https://cra.link/PWA
+precacheAndRoute(self.__WB_MANIFEST);
+
+// Set up App Shell-style routing, so that all navigation requests
+// are fulfilled with your index.html shell. Learn more at
+// https://developers.google.com/web/fundamentals/architecture/app-shell
+const fileExtensionRegexp = new RegExp('/[^/?]+\\.[^/]+$');
+registerRoute(
+  // Return false to exempt requests from being fulfilled by index.html.
+  ({ request, url }: { request: Request; url: URL }) => {
+    // If this isn't a navigation, skip.
+    if (request.mode !== 'navigate') {
+      return false;
+    }
+
+    // If this is a URL that starts with /_, skip.
+    if (url.pathname.startsWith('/_')) {
+      return false;
+    }
+
+    // If this looks like a URL for a resource, because it contains
+    // a file extension, skip.
+    if (url.pathname.match(fileExtensionRegexp)) {
+      return false;
+    }
+
+    // Return true to signal that we want to use the handler.
+    return true;
+  },
+  createHandlerBoundToURL(process.env.PUBLIC_URL + '/index.html')
+);
+
+// This allows the web app to trigger skipWaiting via
+// registration.waiting.postMessage({type: 'SKIP_WAITING'})
+self.addEventListener('message', (event) => {
+  if (event.data && event.data.type === 'SKIP_WAITING') {
+    self.skipWaiting();
+  }
+});
+
+registerRoute(
+  /https:\/\/buyprint\.openstax\.org/,
+  new StaleWhileRevalidate({ // can't use cachefirst because responses are opaque
+    cacheName: 'buy-print',
+    plugins: [
+      new ExpirationPlugin({
+        maxEntries: 20,
+      }),
+    ],
+  })
+);
+
+registerRoute(
+  /https:\/\/cdnjs\.cloudflare\.com\/ajax\/libs\/mathjax\//,
+  new StaleWhileRevalidate({ // can't use cachefirst because responses are opaque
+    cacheName: 'cdn-assets',
+    plugins: [
+      new ExpirationPlugin({
+        maxEntries: 100, // mathjax loads a lot of files
+      }),
+    ],
+  })
+);
+
+registerRoute(
+  /\/cms\/assets\//,
+  new StaleWhileRevalidate({
+    cacheName: 'cms-assets',
+    plugins: [
+      new ExpirationPlugin({
+        maxEntries: 20,
+      }),
+    ],
+  })
+);
+
+registerRoute(
+  /\/contents|resources|extras\//,
```

Comment:
incidentally, i forgot to check if this still works on the rap urls, because they have a prefix on them. does it?

Reply:
yes, it does - I've checked on Firefox and Chrome